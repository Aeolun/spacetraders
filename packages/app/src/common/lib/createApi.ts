import {FleetApi, SystemsApi, DefaultApi, AgentsApi, ContractsApi, FactionsApi, Configuration} from "spacetraders-sdk";
import globalAxios from 'axios';
import axiosRetry from 'axios-retry'

axiosRetry(globalAxios, { retries: 3, retryCondition: (error) => {
  return (error.response?.status && error.response?.status >= 500) || axiosRetry.isNetworkOrIdempotentRequestError(error);
  },
  retryDelay: (retryCount, error) => {
  if (error.response?.status === 429) {
    return error.response.headers['retry-after'] ? error.response.headers['retry-after'] * 1000 : retryCount * 1000;
  } else {
    return retryCount * 1000;
  }
} });

export const requestCounts: Record<string, {
  count: number,
}> = {}
globalAxios.interceptors.response.use(response => {
  if (response.config.url) {
    const url = new URL(response.config.url)
    const logUrl = url.pathname.replace(/PHANTASM-[0-9A-F]+/g, '$').replace(/X1-[A-Z0-9]+-[0-9A-Z]+/g, '$').replace(/X1-[A-Z0-9]+/g, '$')
    if (!requestCounts[logUrl]) {
      requestCounts[logUrl] = {
        count: 0,
      }
    }
    requestCounts[logUrl].count++;
  }
  return response
})

export interface APIInstance {
    systems: SystemsApi,
    fleet: FleetApi,
    default:DefaultApi,
    agents: AgentsApi,
    contracts: ContractsApi,
    factions: FactionsApi
}

export default (agentToken: string, basePath?: string) => {
    const configuration = new Configuration({
        basePath: basePath ?? process.env.API_ENDPOINT,
        accessToken: agentToken,
        
    })
    return {
        systems: new SystemsApi(configuration, undefined, globalAxios),
        fleet: new FleetApi(configuration, undefined, globalAxios),
        default: new DefaultApi(configuration, undefined, globalAxios),
        agents: new AgentsApi(configuration, undefined, globalAxios),
        contracts: new ContractsApi(configuration, undefined, globalAxios),
        factions: new FactionsApi(configuration, undefined, globalAxios)
    }
}