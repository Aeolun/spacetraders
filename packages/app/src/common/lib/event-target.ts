class TypedEventTarget<TP extends Event> extends EventTarget {
  public dispatchEvent(e: Event & TP): boolean {
    return super.dispatchEvent(e);
  }

  public dispatch(e: TP): boolean {
    return this.dispatchEvent(Object.assign(new Event(e.type), e));
  }

  // @ts-expect-error do not feel like fixing this
  public addEventListener<
    T extends TP['type'],
    E extends TP & { type: T }
  >(type: T, listener: ((e: E) => boolean) | null) {
    // @ts-expect-error do not feel like fixing this
    super.addEventListener(type, listener);
  }

  public removeEventListener(type: TP['type']) {
    super.removeEventListener(type, null)
  }
}