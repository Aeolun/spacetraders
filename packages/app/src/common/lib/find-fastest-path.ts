import salesman from 'salesman.js'

class MatrixBuilder {

  public vertexes: {name: string, x: number, y: number}[] = [];

  constructor() {}

  public addVertex(name: string, x: number, y: number) {
    this.vertexes.push({name, x, y});
  }

  private distance(a: {x: number, y: number}, b: {x: number, y: number}) {
    return Math.sqrt(Math.pow(a.x - b.x, 2) + Math.pow(a.y - b.y, 2))
  }

  get size() {
    return this.vertexes.length
  }

  public build() {
    const size = this.vertexes.length
    const matrix = Array(size).fill(0).map(() => Array(size).fill(0));

    for (let i = 0; i < size; i++) {
      const vertexA = this.vertexes[i];
      for (let j = 0; j < size; j++) {
        const vertexB = this.vertexes[j];
        matrix[i][j] = this.distance(vertexA, vertexB);
      }
    }
    return matrix;
  }

  public getNameForIndex(index: number) {
    return this.vertexes[index].name
  }

  public pathToNames(path: number[]) {
    return path.map(i => this.getNameForIndex(i))
  }

  public getBoundaries() {
    const x = this.vertexes.map(i => i.x)
    const y = this.vertexes.map(i => i.y)
    return {
      minX: Math.min(...x),
      maxX: Math.max(...x),
      minY: Math.min(...y),
      maxY: Math.max(...y)
    }
  }
}

export interface Vertex {
  name: string
  x: number
  y: number
}

export const findFastestPath = async (vertices: Vertex[], startNode: string): Promise<{
  path: string[],
  timings: {
    matrixBuildTime: number,
    createSolverTime: number,
    solveTime: number
  }
}> => {
  const matrix = new MatrixBuilder()

  for (const vertex of vertices) {
    matrix.addVertex(vertex.name, vertex.x, vertex.y)
  }

  const startMatrixBuild = Date.now()
  var tspSolverOpts = {
    numNodes: matrix.size,
    costs: matrix.build()
  };
  const matrixBuildTime = Date.now() - startMatrixBuild

  const startSolver = Date.now()

  const depotNode = 0
  var tspSearchOpts = {
    computeTimeLimit: 100,
    depotNode: vertices.findIndex(i => i.name === startNode)
  };
  const createSolverTime = Date.now() - startSolver

  const start = Date.now()

  return new Promise((resolve, reject) => {
    const solution: number[] = salesman.solve(vertices, 0.999995);
    const solveTime = Date.now() - start
    //solution.unshift()
    console.log(solution)




    return resolve({
      path: solution.map(i => vertices[i].name),
      timings: {
        matrixBuildTime,
        createSolverTime,
        solveTime
      }
    })
  });
}