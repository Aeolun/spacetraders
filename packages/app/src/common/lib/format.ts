export const format = Intl.NumberFormat('en-US', { maximumFractionDigits: 0 });