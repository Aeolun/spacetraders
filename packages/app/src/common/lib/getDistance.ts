export const getDistance = (pos1: {x: number, y: number}, pos2: {x: number, y: number}) => {
    return Math.sqrt(Math.pow(pos1.x-pos2.x, 2) + Math.pow(pos1.y-pos2.y, 2))
}