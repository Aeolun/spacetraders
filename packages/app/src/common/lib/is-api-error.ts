import {AxiosError} from "axios";

export interface ApiError {
  error: {
    code: number,
  }
}
export const isApiError = (response: any): response is ApiError => {
  if (typeof response !== "object") {
    return false
  }
  if ('error' in response && 'code' in response.error) {
    return true;
  }
  return false;
}

export const isSTApiErrorResponse = (error: any): error is AxiosError<ApiError> => {
    if (error.isAxiosError && error.response) {
        return isApiError(error.response.data)
    }
    return false
}