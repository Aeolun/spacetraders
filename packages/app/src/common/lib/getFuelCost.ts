import {getDistance} from "@common/lib/getDistance";

export const getFuelCost = (pos1: {x: number, y: number}, pos2: {x: number, y: number}) => {
    return Math.ceil(getDistance(pos1, pos2) / 100) * 220
}