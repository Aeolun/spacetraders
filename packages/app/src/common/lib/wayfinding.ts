import { DijkstraCalculator, LinkedListItem, PathReturnProperties } from '@aeolun/dijkstra-calculator'
import {prisma, System, Waypoint} from '@common/prisma'
import {getDistance} from "@common/lib/getDistance";
import { System as STSystem } from 'spacetraders-sdk'

export interface WayfindingSystem {
    symbol: string
    x: number
    y: number
    hasFuel: boolean
    hasJumpGate: boolean
    jumpConnectedSystems: string[]
}

export interface SimpleSystem {
    symbol: string
    x: number
    y: number
    hasFuel?: boolean
    hasJumpGate?: boolean
}

type WayfindDbSystem = System & { waypoints: ({ symbol: string, type: string, jumpConnectedTo: { symbol: string }[] })[] }

export class Wayfinding {
    dijkstra?: DijkstraCalculator<'fuel'>
    jumpDijkstra?: DijkstraCalculator<'fuel'>
    alreadyConnected = new Set<string>()
    edges = 0
    furthestDistance = 2000
    systemMap: Record<string, {x: number, y: number}> = {}
    waypointMap: Record<string, {x: number, y: number, systemSymbol: string}> = {}
    systemArray: SimpleSystem[] = []
    init: Promise<void>
    connectedSystems: Record<string, {
        system: string,
        distance: number
        methods: string[]
    }[]> = {}
    constructor() {}

    async findRoute(fromSystem: string, toSystem: string, fuel: { max: number, current: number }) {
        await this.init

        if (!this.dijkstra) {
            throw new Error("No dijkstra")
        }
        return this.dijkstra.calculateShortestPathAsLinkedListResult(fromSystem, toSystem, {
            supplies: {
                fuel: fuel.current
            },
            supplyCapacity: {
                fuel: fuel.max
            }
        })
    }

    async findJumpRoute(fromSystem: string, toSystem: string, fuel: { max: number, current: number }) {
        await this.init

        if (!this.jumpDijkstra) {
            throw new Error("No jump dijkstra")
        }
        return this.jumpDijkstra.calculateShortestPathAsLinkedListResult(fromSystem, toSystem, {
            supplies: {
                fuel: fuel.current
            },
            supplyCapacity: {
                fuel: fuel.max
            }
        })
    }

    addConnection(fromSystem: string, toSystem: string, distance: number, methods: string[]) {
        if (!this.connectedSystems[fromSystem]) {
            this.connectedSystems[fromSystem] = [{
                system: toSystem,
                distance,
                methods
            }]
        } else if (!this.connectedSystems[fromSystem].some(connection => connection.system === toSystem)) {
            // if not connected yet, add toSystem
            this.connectedSystems[fromSystem].push({
                system: toSystem,
                distance,
                methods
            })
        } else if(this.connectedSystems[fromSystem].find(connection => connection.system === toSystem)) {
            // if already connected, add methods
            this.connectedSystems[fromSystem].find(connection => connection.system === toSystem)?.methods.push(...methods)
        }
    }

    getConnectionsForSystem(system: string) {
        return this.connectedSystems[system]
    }

    addRoutes(system: WayfindingSystem) {
        if (system.hasJumpGate) {
            system.jumpConnectedSystems.forEach(waypointSymbol => {
                const connectionId = [system.symbol, waypointSymbol].sort().join('-')+'jump'
                if (this.alreadyConnected.has(connectionId)) {
                    return;
                }
                this.alreadyConnected.add(connectionId)

                const systemSymbol = system.symbol
                if (!this.waypointMap[waypointSymbol]) {
                    throw new Error(`Cannot find waypoint ${waypointSymbol} in waypointMap`)
                }
                const toSystemSymbol = this.waypointMap[waypointSymbol].systemSymbol

                this.dijkstra.addEdge(systemSymbol, toSystemSymbol, {
                    id: 'jump',
                    weight: getDistance(system, this.waypointMap[waypointSymbol]),
                })
                this.jumpDijkstra.addEdge(systemSymbol, toSystemSymbol, {
                    id: 'jump',
                    weight: getDistance(system, this.waypointMap[waypointSymbol])
                })
                this.addConnection(systemSymbol, toSystemSymbol, getDistance(system, this.waypointMap[waypointSymbol]), ['jump'])
                this.edges++
            })
        }
        const nearbySystems = this.systemArray.filter(nearbySystem => {
            return getDistance(system, nearbySystem) <= this.furthestDistance+1 && !nearbySystem.hasJumpGate
        }).map(nearbySystem => {
            return {
                system: nearbySystem,
                distance: getDistance(system, nearbySystem)
            }
        }).sort((a, b) => {
            // closest 5 systems
            return a.distance - b.distance
        }).slice(0, 7)
        // nearby systems only counts the closest 5 systems that do NOT have a jump gate for purposes of adding edges
        nearbySystems.forEach(nearbySystem => {
            const connectionId = [system.symbol, nearbySystem.system.symbol].sort().join('-')
            if (this.alreadyConnected.has(connectionId)) {
                return;
            }
            this.alreadyConnected.add(connectionId)
            if (nearbySystem.distance > 2500) {
                // increase cost function by fuel cost
                this.dijkstra.addEdge(system.symbol, nearbySystem.system.symbol, {
                    weight: nearbySystem.distance * 1.2,
                    // weight: ((Math.round(nearbySystem.distance) * (30 / 30) + 15) * 10) + (nearbySystem.distance / 100 * 120),
                    id: 'cruise',
                    consumes: {
                        fuel: nearbySystem.distance
                    }
                })
                this.edges++
                this.dijkstra.addEdge(system.symbol, nearbySystem.system.symbol, {
                    weight: nearbySystem.distance * 2,
                    // weight: ((Math.round(nearbySystem.distance) * (15 / 30) + 15) * 10) + (nearbySystem.distance * 2 / 100 * 120),
                    id: 'burn',
                    consumes: {
                        fuel: nearbySystem.distance * 2
                    }
                })
                this.edges++
                this.dijkstra.addEdge(system.symbol, nearbySystem.system.symbol, {
                    weight: nearbySystem.distance * 100,
                    // weight: ((Math.round(nearbySystem.distance) * (300 / 30) + 15) * 10) + (1 / 100 * 120),
                    id: 'drift',
                    consumes: {
                        fuel: 1
                    }
                })
                this.edges++

                this.addConnection(system.symbol, nearbySystem.system.symbol, nearbySystem.distance, ['cruise', 'burn', 'drift'])
            }
        })
    }

    resetData(systems: SimpleSystem[]) {
        this.systemMap = {}
        this.waypointMap = {}
        this.systemArray = []

        systems.forEach(system => {
            this.systemMap[system.symbol] = {
                x: system.x,
                y: system.y
            }
            this.systemArray.push({
                symbol: system.symbol,
                x: system.x,
                y: system.y
            })
        });


        this.alreadyConnected = new Set<string>()
        this.edges = 0;
        this.furthestDistance = 0;
        this.dijkstra = new DijkstraCalculator((current, target) => {
            //return 0
            return getDistance(this.systemMap[current], this.systemMap[target]) * 0.999;
        })
        this.jumpDijkstra = new DijkstraCalculator((current, target) => {
            //return 0
            return getDistance(this.systemMap[current], this.systemMap[target]) * 0.999;
        })

        systems.forEach(system => {
            this.dijkstra.addVertex(system.symbol, {
                recover: system.hasFuel ? {
                    fuel: true
                } : undefined
            })
            this.jumpDijkstra.addVertex(system.symbol, {
                recover: system.hasFuel ? {
                    fuel: true
                } : undefined
            })
        })
    }

    async loadWaypointsFromJson(systems: STSystem[]) {
        const simpleSystems = systems.map(system => {
            return {
                symbol: system.symbol,
                x: system.x,
                y: system.y,
                hasFuel: system.waypoints.some(wp => wp.type === 'PLANET' || wp.type === 'MOON'),
                hasJumpGate: system.waypoints.some(wp => wp.type === 'JUMP_GATE'),
                jumpGateWaypoint: system.waypoints.find(wp => wp.type === 'JUMP_GATE')?.symbol
            }
        })

        this.resetData(simpleSystems);

        const {furthest, furthestDistance} = findClosestSystems(simpleSystems)
        this.furthestDistance = furthestDistance
        console.log("furthest distance between two systems", furthestDistance, ' between ', furthest)

        systems.forEach(system => {
            system.waypoints.forEach(wp => this.waypointMap[wp.symbol] = {x: system.x, y: system.y, systemSymbol: system.symbol })
        });
        systems.forEach(system => {
            this.addRoutes({
                symbol: system.symbol,
                x: system.x,
                y: system.y,
                hasJumpGate: system.waypoints.some(wp => wp.type === 'JUMP_GATE'),
                // this is obviously incorrect, but it's a decent approximation
                hasFuel: system.waypoints.some(wp => wp.type === 'PLANET' || wp.type === 'MOON'),
                jumpConnectedSystems: system.waypoints.some(wp => wp.type === 'JUMP_GATE') ? simpleSystems.filter(s => s.symbol !== system.symbol && s.hasJumpGate && getDistance(system, s) <= 2500).map(s => s.jumpGateWaypoint) : []
            })
        })

        console.log(`Reloaded edges for ${systems.length} systems, ${this.edges} edges`)
    }

    async addSystemFromDatabase(systemSymbol: string) {
        const system = await prisma.system.findFirst({
            include: {
                waypoints: {
                    where: {
                        type: 'JUMP_GATE'
                    },
                    include: {
                        jumpConnectedTo: {
                            select: {
                                symbol: true
                            }
                        }
                    }
                }
            },
            where: {
                symbol: systemSymbol
            }
        })
        this.addRoutes(this.systemToWayfindingSystem(system))
    }


    private systemToWayfindingSystem(system: WayfindDbSystem) {
        const jumpTargets: string[] = []

        const jumpGate = system.waypoints.find(wp => wp.type === 'JUMP_GATE')
        jumpGate?.jumpConnectedTo.forEach(jt => {
            jumpTargets.push(jt.symbol)
        })
        return {
            symbol: system.symbol,
            x: system.x,
            y: system.y,
            hasFuel: system.hasFuel,
            hasJumpGate: system.hasJumpGate,
            jumpConnectedSystems: jumpTargets
        }
    }

    async loadWaypointsFromDb() {
        const systems = await prisma.system.findMany({
            include: {
                waypoints: {
                    where: {
                        type: 'JUMP_GATE'
                    },
                    select: {
                        symbol: true,
                        jumpConnectedTo: {
                            select: {
                                symbol: true
                            }
                        }
                    }
                }
            },
            where: {
                waypoints: {
                    some: {
                        symbol: {
                            contains: 'X1'
                        }
                    }
                }
            }
        })
        console.log("loaded systems", systems.length)

        const wayfindingSystems = systems.map(system => {
            return this.systemToWayfindingSystem(system)
        });

        this.resetData(wayfindingSystems);

        systems.forEach(system => {
            system.waypoints.forEach(wp => {
                this.waypointMap[wp.symbol] = {x: system.x, y: system.y, systemSymbol: system.symbol }
            })
        });

        const {furthest, furthestDistance} = findClosestSystems(wayfindingSystems)
        this.furthestDistance = furthestDistance
        console.log("furthest distance between two systems", furthestDistance, ' between ', furthest)

        wayfindingSystems.forEach(system => {
            this.addRoutes(system)
        })

        console.log(`Reloaded edges for ${systems.length} systems, ${this.edges} edges`)
    }


}

export function findClosestSystems(systems: SimpleSystem[]) {
    let furthestDistance = 0
    let furthest = ''
    systems.forEach(system => {
        let closest;
        let closestSystem = 100000000
        systems.forEach(nearbySystem => {
            if (system.symbol !== nearbySystem.symbol) {
                const distance = getDistance(system, nearbySystem)
                if (distance <= closestSystem) {
                    closestSystem = distance
                    closest = nearbySystem.symbol
                }
            }
        })

        if (closestSystem > furthestDistance) {
            furthestDistance = closestSystem
            furthest = system.symbol + '-' + closest
        }
    })
    return {
        furthest,
        furthestDistance
    }
}