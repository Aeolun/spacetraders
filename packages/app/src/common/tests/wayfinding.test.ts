import {Wayfinding} from "@common/lib/wayfinding";
import systems from './mocks/systems.json'
import { expect, test } from "vitest";
import {System} from "spacetraders-sdk";
import {beforeAll} from "vitest";

let wayfinder: Wayfinding
test('wayfinding', async () => {
  expect(1).toBe(1);
});
// beforeAll(async () => {
//   wayfinder = new Wayfinding()
//   await wayfinder.loadWaypointsFromJson(systems as System[])
// })
// test('wayfinding', async () => {
//   const route = await wayfinder.findRoute('X1-SD19', 'X1-GY82', {
//     max: 1000,
//     current: 500
//   })
//   expect(route.finalPath).toMatchSnapshot()
// })
//
// test('wayfinding 2', async () => {
//   const route = await wayfinder.findRoute('X1-CH98', 'X1-MF76', {
//     max: 1000,
//     current: 500
//   })
//   expect(route.finalPath).toMatchSnapshot()
//   expect(wayfinder.getConnectionsForSystem('X1-MF76')).toMatchSnapshot()
//   expect(wayfinder.getConnectionsForSystem('X1-HU19')).toMatchSnapshot()
// })