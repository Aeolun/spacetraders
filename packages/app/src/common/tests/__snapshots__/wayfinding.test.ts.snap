// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wayfinding 1`] = `
[
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 500,
    },
    "source": "X1-SD19",
    "target": "X1-ZX54",
    "weight": 2339.076954698156,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-ZX54",
    "target": "X1-SS71",
    "weight": 2479.5132586860673,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-SS71",
    "target": "X1-JD5",
    "weight": 2295.6624316305742,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-JD5",
    "target": "X1-RD42",
    "weight": 1778.5648709001312,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-RD42",
    "target": "X1-NR91",
    "weight": 2364.1541827892697,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-NR91",
    "target": "X1-TJ93",
    "weight": 2465.434647278244,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-TJ93",
    "target": "X1-FM88",
    "weight": 2398.216003616021,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-FM88",
    "target": "X1-CH98",
    "weight": 2198.7398663780123,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-CH98",
    "target": "X1-GY82",
    "weight": 819.0579955045919,
  },
]
`;

exports[`wayfinding 2 1`] = `
[
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 500,
    },
    "source": "X1-CH98",
    "target": "X1-QG81",
    "weight": 2292.4720281826776,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {},
    "source": "X1-QG81",
    "target": "X1-TF20",
    "weight": 2142.4658690396914,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-TF20",
    "target": "X1-YP11",
    "weight": 2136.0978442009628,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-YP11",
    "target": "X1-AJ69",
    "weight": 1915.1067855344254,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {},
    "source": "X1-AJ69",
    "target": "X1-HF79",
    "weight": 2469.4705910376824,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-HF79",
    "target": "X1-DP76",
    "weight": 2252.115893998353,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-DP76",
    "target": "X1-AF13",
    "weight": 2004.1619196062977,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-AF13",
    "target": "X1-XS56",
    "weight": 1731.0577691111293,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-XS56",
    "target": "X1-PP4",
    "weight": 1878.3588581525096,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-PP4",
    "target": "X1-RJ56",
    "weight": 1804.7861369148422,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-RJ56",
    "target": "X1-GF33",
    "weight": 2122.507950515145,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-GF33",
    "target": "X1-SC40",
    "weight": 2102.1048974777636,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-SC40",
    "target": "X1-HX33",
    "weight": 2067.8587959529536,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-HX33",
    "target": "X1-MA87",
    "weight": 1454.5683208429916,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-MA87",
    "target": "X1-CD26",
    "weight": 1825.260803282643,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-CD26",
    "target": "X1-RM3",
    "weight": 1851.4623949732277,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-RM3",
    "target": "X1-AS7",
    "weight": 2427.800032951643,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-AS7",
    "target": "X1-QR18",
    "weight": 2429.4742229544236,
  },
  {
    "consumes": {
      "fuel": 1,
    },
    "edge": "drift",
    "recover": {
      "fuel": 1,
    },
    "source": "X1-QR18",
    "target": "X1-DJ77",
    "weight": 272751.84692316933,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-DJ77",
    "target": "X1-DZ78",
    "weight": 1464.3285150539136,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {},
    "source": "X1-DZ78",
    "target": "X1-UK6",
    "weight": 1153.2185395665472,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-UK6",
    "target": "X1-AZ63",
    "weight": 2220.636170109818,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-AZ63",
    "target": "X1-SZ67",
    "weight": 1451.0933808683712,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-SZ67",
    "target": "X1-ZX45",
    "weight": 2386.5223652838454,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-ZX45",
    "target": "X1-MF94",
    "weight": 2397.705569914705,
  },
  {
    "consumes": undefined,
    "edge": "jump",
    "recover": {
      "fuel": 0,
    },
    "source": "X1-MF94",
    "target": "X1-VK68",
    "weight": 1816.1313278504945,
  },
  {
    "consumes": {
      "fuel": 1,
    },
    "edge": "drift",
    "recover": {
      "fuel": 1,
    },
    "source": "X1-VK68",
    "target": "X1-MF76",
    "weight": 278140.414179601,
  },
]
`;

exports[`wayfinding 2 2`] = `
[
  {
    "distance": 2781.40414179601,
    "methods": [
      "cruise",
      "burn",
      "drift",
    ],
    "system": "X1-VK68",
  },
  {
    "distance": 3448.37193469614,
    "methods": [
      "cruise",
      "burn",
      "drift",
    ],
    "system": "X1-U97",
  },
  {
    "distance": 3505.5889662081036,
    "methods": [
      "cruise",
      "burn",
      "drift",
    ],
    "system": "X1-YV95",
  },
]
`;

exports[`wayfinding 2 3`] = `
[
  {
    "distance": 8751.247968147172,
    "methods": [
      "cruise",
      "burn",
      "drift",
    ],
    "system": "X1-MF76",
  },
]
`;
